### PR TITLE
Upgraded SnapStart pattern to SpringBoot 3 and Java 17

### PR DESCRIPTION
**Pull request by iamkhattar**
*Issue #, if available:*

*Description of changes:*
-> Upgraded the SnapStart pattern to leverage Java 17 and SpringBoot 3.
-> Added tests for the repo including integration tests that leverage TestContainers with DynamoDB Local.
-> Added Powertools for observability
-> Changed DynamoDB Client to enhanced to be in line with the sam templates: https://github.com/aws-samples/java-crud-microservice-template/tree/main

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by iamkhattar** [comment]
@msailes here is the PR for the Java 17 Upgrade to the SnapStart template

File: apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/StreamLambdaHandler.java [hunk 1]
```diff
@@ -7,13 +7,20 @@
...
+import org.apache.logging.log4j.Logger;
+import software.amazon.lambda.powertools.logging.Logging;
+import software.amazon.lambda.powertools.metrics.Metrics;
+import software.amazon.lambda.powertools.tracing.Tracing;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 
 public class StreamLambdaHandler implements RequestStreamHandler {
```
Comment (olegz) [inline apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/StreamLambdaHandler.java:20]: Also, once M2 is released we won't need this class at all and instead use the new `SpringDelegatingLambdaContainerHandler` as demonstrated here https://github.com/awslabs/aws-serverless-java-container/tree/main/samples/springboot3/alt-pet-store

File: apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/config/DeserializationConfig.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class DeserializationConfig {
+
+    @Bean
+    public ObjectMapper objectMapper() {
```
[the remark this answers is not in the document]
Comment (iamkhattar) [reply apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/config/DeserializationConfig.java]: Thanks for the link, this was great. I have removed this class considering this is already default (totally my bad here)

File: apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/StreamLambdaHandler.java (hunk 1 above)
Comment (iamkhattar) [reply apigw-lambda-snapstart/UnicornFunction/src/main/java/com/unicorn/store/StreamLambdaHandler.java:20]: I would be happy to raise a pr to to update this when m2 is out

**Review by msailes — APPROVED** [review]
LGTM!

**Comment by iamkhattar** [comment]
@olegz thanks for the review, I have removed the deserialization configuration now and will be happy to pr the update to template when m2 is out

**Pull request merged**